circuits/update_state_tree.circom

```diff
@@ -1,106 +1,266 @@
+  signal input msg_tree_path_index[depth];
+
+  // State tree
+  signal input state_tree_leaf;
+  signal private input state_tree_data[state_tree_data_length];
+
+  signal input state_tree_max_leaf_index;
   signal input state_tree_root;
-  signal private input state_tree_path_elements[levels];
-  signal private input state_tree_path_index[levels];
-
-  // Length of the encrypted data
-  var encrypted_data_length = 7;
-
-  // Length of the message (without signature, decrypted)
-  var message_length = 3;
-
-  // NOTE: Last 3 elements in the arr
-  // MUST BE THE SIGNATURE!
-  /*
-      [0] - iv (generated when msg is encrypted)
-      [1] - publickey_x
-      [2] - publickey_y
-      [3] - action
-      [4] - signature_r8x
-      [5] - signature_r8y
-      [6] - signature_s
-   */
-  signal input encrypted_data[encrypted_data_length];
-
-  // Shared private key to decrypt encrypted data
+  signal private input state_tree_path_elements[depth];
+  signal private input state_tree_path_index[depth];
+
+  // Random leaf (0th index on state tree)
+  signal input random_leaf;
+  signal private input random_leaf_path_elements[depth];
+  signal private input random_leaf_path_index[depth];
+
+  // Shared keys
   signal private input ecdh_private_key;
+  signal input ecdh_public_key[2];
+
+  // Logic parameters
+  signal private input no_op;
+
+  // Check 1. Coordinator is using correct private key
+  component derived_pub_key = PublicKey();
+  derived_pub_key.private_key <== ecdh_private_key;
+
+  derived_pub_key.public_key[0] === coordinator_public_key[0]
+  derived_pub_key.public_key[1] === coordinator_public_key[1]
 
-  // Inputs that are currently in the state tree
-  // that wants to be updated
-  signal private input existing_public_key[2];
-  signal private input existing_state_tree_leaf;
+  component ecdh = Ecdh();
+  ecdh.private_key <== ecdh_private_key;
+  ecdh.public_key[0] <== ecdh_public_key[0];
+  ecdh.public_key[1] <== ecdh_public_key[1];
 
-  // Construct leaf values
-  component encrypted_data_hash = Hasher(encrypted_data_length);
-  encrypted_data_hash.key <== 0;
-  for (var i = 0; i < encrypted_data_length; i++) {
-    encrypted_data_hash.in[i] <== encrypted_data[i];
+  // Check 2. Assert decrypted messages are the same
+  component decrypted_command = Decrypt(message_length - 1);
+  decrypted_command.private_key <== ecdh.shared_key;
+  for (var i = 0; i < message_length; i++) {
+    decrypted_command.message[i] <== message[i];
   }
 
-  // **** 1. Make sure the leaf exists in the cmd tree **** //
-  component cmd_tree_value_exists = LeafExists(levels);
-  cmd_tree_value_exists.root <== cmd_tree_root;
-  cmd_tree_value_exists.leaf <== encrypted_data_hash.hash;
-  for (var i = 0; i < levels; i++) {
-    cmd_tree_value_exists.path_elements[i] <== cmd_tree_path_elements[i];
-    cmd_tree_value_exists.path_index[i] <== cmd_tree_path_index[i];
+  for (var i = 0; i < message_length - 1; i++) {
+    decrypted_command.out[i] === command[i];
```

Comment:
Style question - do we need the extra variable `command` and the extra constraints it entails? Can't we use decrypted command directly?

Reply:
@kobigurk we definitely can, the specs seems to want it tho. Maybe it can help double check if the coordinator managed to get similar decryption results to the circuit?

Reply:
I don't think it adds a double check, though let's chat about it :)